examples/07_end_to_end/wikimedia.ipynb

```diff
@@ -0,0 +1,464 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5667cea7-43c5-471c-9859-757c6ac18be2",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "import json\n",
+    "import threading\n",
+    "\n",
+    "from sseclient import SSEClient as EventSource\n",
+    "\n",
+    "import csp\n",
+    "from csp.impl.pushadapter import PushInputAdapter\n",
+    "from csp.impl.wiring import py_push_adapter_def\n",
+    "\n",
+    "class WikiData(csp.Struct):\n",
+    "    servername: str\n",
+    "    servername_count: int\n",
+    "    timestamp: str\n",
+    "    event: str\n",
+    "\n",
+    "# Define the runtime implementation of our adapter\n",
+    "class FetchWikiDataAdapter(PushInputAdapter):\n",
+    "    def __init__(self, url: str):\n",
+    "        self._thread = None\n",
+    "        self._running = False\n",
+    "        self._url = url\n",
+    "\n",
+    "    def start(self, starttime, endtime):\n",
+    "        print(\"FetchWikiDataAdapter::start\")\n",
+    "        self._source = EventSource(self._url)\n",
+    "        self._running = True\n",
+    "        self._thread = threading.Thread(target=self._run)\n",
+    "        self._thread.start()\n",
+    "\n",
+    "    def stop(self):\n",
+    "        print(\"FetchWikiDataAdapter::stop\")\n",
+    "        if self._running:\n",
+    "            self._running = False\n",
+    "            self._thread.join()\n",
+    "            self._source.close()\n",
+    "\n",
+    "    def _run(self):\n",
+    "        servernames = dict([])\n",
+    "        while self._running:\n",
+    "            for item in self._source:\n",
+    "                if item.event == 'message':\n",
+    "                    try:\n",
+    "                        change = json.loads(item.data)\n",
+    "                    except ValueError:\n",
+    "                        pass\n",
+    "                    else:\n",
+    "                        # discard canary events\n",
+    "                        # WMF Data Engineering team produces artificial 'canary' events into \n",
+    "                        # each stream multiple times an hour. The presence of these canary\n",
+    "                        # events in a stream allow us to differentiate between a broken event\n",
+    "                        # stream, and an empty one.\n",
+    "                        # We will also filter bot-generated events\n",
+    "                        if change['meta']['domain'] == 'canary' or re.search('bot', change['user'], re.IGNORECASE):\n",
+    "                            continue\n",
+    "                        timestamp = change['meta']['dt']\n",
+    "                        event = f\"{timestamp}:: {change['user']} edited {change['title']}\"\n",
+    "                        servername = change['server_name']\n",
```

Comment:
This is looking good.  One last point.  The WikiData struct should be the same for both realtime / historical.  This is the strongest point of CSP, that APIs are defined by timeseries of the same type where you can easily replace realtime and historical edges and everything else works the same.
I recommend making this last change:
- Define the struct once, you can leave server_count in there
- remove the logic of setting count in the realtime adapter, it really shouldnt be doing anything other than pushing data
- write another node like "compute_server_count" that takes the raw struct in, creates a copy with the count set and ticks it out
- if possible, write one graph that takes ts[WikiData] in and then show how you can pass either the realtime edge or the historical edge into it.  It can be something as trivial as calling compute_server_count and passing that to pretty_print

Reply:
I think I got it, let me know if there's anything else.